whistl/classifier.py

```diff
@@ -350,6 +354,181 @@ def train_with_erm(classifier, map_file, train_dirs, tune_dirs, gene_file, num_e
+
+                    loss_function = nn.BCEWithLogitsLoss(weight=batch_weights)
+                    optimizer.zero_grad()
+                    output = task_head(representation(expression))
+                    loss = loss_function(output, labels)
+                    train_loss += float(loss)
+
+                    train_correct += util.count_correct(output, labels)
+
+                    loss.backward()
+                    optimizer.step()
+
+                # Disable the gradient and switch into model evaluation mode
+                with torch.no_grad():
+                    representation.eval()
+                    task_head.eval()
+
+                    tune_loss = 0
+                    tune_correct = 0
+                    for tune_batch in tune_loader:
+                        expression, labels, ids = tune_batch
+                        expression = expression.to(device)
+                        tune_labels = labels.to(device).double()
+
+                        loss_function = nn.BCEWithLogitsLoss()
+
+                        tune_output = task_head(representation(expression))
+
+                        loss = loss_function(tune_output, tune_labels)
+                        tune_loss += float(loss)
+                        tune_correct += util.count_correct(tune_output, tune_labels)
+
+                    # Save the model
+                    if save_file_dir is not None:
+                        if best_tune_loss is None or tune_loss < best_tune_loss:
+                            best_tune_loss = tune_loss
+
+                            head_file_name = 'erm_{}_head.pkl'.format(task)
+                            head_file_path = os.path.join(save_file_dir, head_file_name)
+                            torch.save(task_head, head_file_path)
+
+                train_accuracy = train_correct / len(train_dataset)
+                tune_accuracy = tune_correct / len(tune_dataset)
+
+                if logger is not None:
+                    logger.info('Epoch {}'.format(epoch))
+                    logger.info('Train loss: {}'.format(train_loss / len(train_dataset)))
+                    logger.info('Tune loss: {}'.format(tune_loss / len(tune_dataset)))
+                    logger.info('Train accuracy: {}'.format(train_accuracy))
+                    logger.info('Tune accuracy: {}'.format(tune_accuracy))
+                    logger.info('Baseline accuracy: {}'.format(baseline))
+
+                results['train_loss'].append(train_loss / len(train_dataset))
+                results['tune_loss'].append(tune_loss / len(tune_dataset))
+                results['train_acc'].append(train_accuracy)
+                results['tune_acc'].append(tune_accuracy)
+        except Exception as e:
+            # Print error
+            logger.error(e, exc_info=True)
+        finally:
+            if save_file_dir is not None:
+                representation_file = os.path.join(save_file_dir, 'erm_representation.pkl')
```

Comment:
Maybe I'm just misunderstanding what you're trying to do here, but is there any reason why you're saving the "best" head model above on line 499 (essentially doing early stopping for the head model), but you're saving the representation from the last epoch here even if it doesn't have the best loss?

It seems to me like you'd also want to save the representation with the best loss (i.e. the representation that corresponds to the saved head model), but I could be confused about the purpose of this.

Reply:
I don't really know what works best yet. The purpose of this commit was to sketch out the base function to use as reference so I can benchmark different methods later. Currently each task is being trained in its entirety before moving onto the next, but it could be done a number of different ways. In a future PR I plan on comparing
- The current method
- Train only one batch per task at a time
- Do multiclass learning instead of multitask (easier training code, probably better performance, but harder to swap out task heads further down the line)

I think the correct way to do it would be to train the representation, without keeping the heads, then freeze the representation weights and retrain the each head on the optimal representation.

All of that is to say that you are correct that the current method doesn't make much sense. I'm alright with that because it's more of a prototype than a method I intend to use in analyses